corehq/ex-submodules/pillowtop/processors/elastic.py

```diff
@@ -159,25 +161,28 @@ def process_changes_chunk(self, changes_chunk):
         return retry_changes, error_changes
 
 
-def send_to_elasticsearch(alias, doc_type, doc_id, es_getter, name, data=None,
-                          retries=MAX_RETRIES, propagate_failure=settings.UNIT_TESTING,
-                          update=False, delete=False, es_merge_update=False):
+def send_to_elasticsearch(index_info, doc_type, doc_id, es_getter, name, data=None,
+                          delete=False, es_merge_update=False):
     """
     More fault tolerant es.put method
     kwargs:
         es_merge_update: Set this to True to use Elasticsearch.update instead of Elasticsearch.index
             which merges existing ES doc and current update. If this is set to False, the doc will be replaced
 
     """
+    alias = index_info.alias
     data = data if data is not None else {}
     current_tries = 0
     es_interface = ElasticsearchInterface(es_getter())
-    retries = 1 if settings.UNIT_TESTING else retries
+    retries = 1 if settings.UNIT_TESTING else MAX_RETRIES
+    propagate_failure = settings.UNIT_TESTING
     while current_tries < retries:
         try:
+            doc_exists = es_interface.doc_exists(alias, doc_id, doc_type)
             if delete:
-                es_interface.delete_doc(alias, doc_type, doc_id)
-            elif update:
+                if doc_exists:
+                    es_interface.delete_doc(alias, doc_type, doc_id)
+            elif doc_exists:
                 params = {'retry_on_conflict': 2}
                 if es_merge_update:
                     es_interface.update_doc_fields(alias, doc_type, doc_id, fields=data, params=params)
```

Comment:
another thought for future:

The only difference between `create` and `index` is that create will raise an error if the doc exists. I don't think we depend on that anywhere so we could just use the `index` method for both create and update (unless we need to 'update_fields'). If we did that (and my other thought above) we wouldn't need to check for doc existence at all except in cases where we want to only do a `fields` update.

Furthermore, in the case of `es_merge_update` we are assuming that the doc already exists so we could just call `update` without the check. If the doc doesn't exist we get a 404 which we can handle. The current functionality will create a new doc with only the fields specified which I think is incorrect behaviour.